File: src/engine/resolution.ts
// resolution.ts — Resolution Refutation Inference Engine

import type { Clause, Literal } from './cnf';
import { clauseKey, negateLiteral } from './cnf';
import type { AskResult } from './kb';

/**
 * Resolution Refutation
 *
 * To prove `goal` (e.g. "¬P_2_1"):
 *  1. Add the negation of the goal as a unit clause: [negateLiteral(goal)]
 *     e.g., to prove ¬P_2_1 → add [P_2_1]
 *  2. Repeatedly pick clause pairs (Ci, Cj) where Ci contains literal L and
 *     Cj contains ¬L.
 *  3. Resolvent = (Ci ∪ Cj) \ {L, ¬L}
 *  4. If resolvent == {} → contradiction → goal is proved.
 *  5. If no new resolvents can be generated → not provable.
 *
 * Each unique resolvent generation counts as +1 inference step.
 */
export function resolve(kbClauses: Clause[], goal: Literal): AskResult {
  // Working set of clauses (KB + negated goal)
  const negGoal: Clause = [negateLiteral(goal)];
  const working: Clause[] = [...kbClauses, negGoal];

  const seenKeys = new Set<string>(working.map(clauseKey));
  const log: string[] = [
    `Goal: prove ${goal}`,
    `Negated goal added: [${negGoal.join(', ')}]`,
    `Initial clause count: ${working.length}`,
  ];

  let inferenceSteps = 0;
  let newClausesFound = true;

  while (newClausesFound) {
    newClausesFound = false;
    const n = working.length;

    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        const resolvents = resolvePair(working[i], working[j]);

        for (const resolvent of resolvents) {
          inferenceSteps++;
          const key = clauseKey(resolvent);

          // Contradiction found — return success
          if (resolvent.length === 0) {
            log.push(
              `Step ${inferenceSteps}: Resolve [${working[i].join(', ')}] ` +
              `with [${working[j].join(', ')}] → ∅  ← CONTRADICTION`
            );
            log.push(`✓ Proved: ${goal}`);
            return { proved: true, inferenceSteps, log };
          }

          if (!seenKeys.has(key)) {
            seenKeys.add(key);
            working.push(resolvent);
            newClausesFound = true;
            log.push(
              `Step ${inferenceSteps}: Resolve [${working[i].join(', ')}] ` +
              `with [${working[j].join(', ')}] → [${resolvent.join(', ')}]`
            );
          }
        }
      }
    }
  }

  log.push(`✗ Could not prove: ${goal} (no contradiction found)`);
  return { proved: false, inferenceSteps, log };
}

/**
 * Given two clauses, find all literals L in Ci and ¬L in Cj,
 * and return all corresponding resolvents.
 * Returns [] if no complementary literal pairs exist.
 */
function resolvePair(ci: Clause, cj: Clause): Clause[] {
  const resolvents: Clause[] = [];

  for (const lit of ci) {
    const complement = negateLiteral(lit);
    if (cj.includes(complement)) {
      // Remove lit from ci and complement from cj
      const resolvent: Clause = [
        ...ci.filter(l => l !== lit),
        ...cj.filter(l => l !== complement),
      ];
      // Deduplicate literals within the resolvent
      const deduped = [...new Set(resolvent)];
      // Skip tautologies (e.g., contains X and ¬X)
      if (!isTautology(deduped)) {
        resolvents.push(deduped);
      }
    }
  }

  return resolvents;
}

/** A clause is a tautology if it contains both L and ¬L. */
function isTautology(clause: Clause): boolean {
  const litSet = new Set(clause);
  return clause.some(lit => litSet.has(negateLiteral(lit)));
}

File: src/python_engine/resolution.py
from typing import List, Set, Tuple
from .cnf import Clause, Literal, negate_literal, clause_key

class AskResult:
    def __init__(self, proved: bool, inference_steps: int, log: List[str]):
        self.proved = proved
        self.inference_steps = inference_steps
        self.log = log

def is_tautology(clause: Clause) -> bool:
    lit_set = set(clause)
    return any(negate_literal(lit) in lit_set for lit in clause)

def resolve_pair(ci: Clause, cj: Clause) -> List[Clause]:
    resolvents = []
    
    for lit in ci:
        complement = negate_literal(lit)
        if complement in cj:
            resolvent = [l for l in ci if l != lit] + [l for l in cj if l != complement]
            deduped = list(dict.fromkeys(resolvent)) # deduplicate while preserving order
            
            if not is_tautology(deduped):
                resolvents.append(deduped)
                
    return resolvents

def resolve(kb_clauses: List[Clause], goal: Literal) -> AskResult:
    neg_goal = [negate_literal(goal)]
    working = list(kb_clauses) + [neg_goal]
    
    seen_keys = {clause_key(c) for c in working}
    
    log = [
        f"Goal: prove {goal}",
        f"Negated goal added: [{', '.join(neg_goal)}]",
        f"Initial clause count: {len(working)}"
    ]
    
    inference_steps = 0
    new_clauses_found = True
    
    while new_clauses_found:
        new_clauses_found = False
        n = len(working)
        
        for i in range(n):
            for j in range(i + 1, n):
                resolvents = resolve_pair(working[i], working[j])
                
                for resolvent in resolvents:
                    inference_steps += 1
                    key = clause_key(resolvent)
                    
                    if len(resolvent) == 0:
                        log.append(f"Step {inference_steps}: Resolve [{', '.join(working[i])}] with [{', '.join(working[j])}] → ∅  ← CONTRADICTION")
                        log.append(f"✓ Proved: {goal}")
                        return AskResult(proved=True, inference_steps=inference_steps, log=log)
                        
                    if key not in seen_keys:
                        seen_keys.add(key)
                        working.append(resolvent)
                        new_clauses_found = True
                        log.append(f"Step {inference_steps}: Resolve [{', '.join(working[i])}] with [{', '.join(working[j])}] → [{', '.join(resolvent)}]")
                        
    log.append(f"✗ Could not prove: {goal} (no contradiction found)")
    return AskResult(proved=False, inference_steps=inference_steps, log=log)
